nano/node/socket.cpp

```diff
@@ -238,55 +238,110 @@ void nano::server_socket::on_connection (std::function<bool (std::shared_ptr<nan
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
+				boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
+					this_l->on_connection (callback_a);
+				}));
+				return;
+			}
 
-						// If the callback returns true, keep accepting new connections
-						if (callback_a (new_connection, ec_a))
-						{
-							this_l->on_connection (callback_a);
-						}
-						else
-						{
-							this_l->node.logger.try_log ("Stopping to accept connections");
-						}
-					}
-					else
-					{
-						this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
-						boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
-							this_l->on_connection (callback_a);
-						}));
-					}
+			if (!ec_a)
+			{
+				// Make sure the new connection doesn't idle. Note that in most cases, the callback is going to start
+				// an IO operation immediately, which will start a timer.
+				new_connection->checkup ();
+				new_connection->start_timer (this_l->node.network_params.network.is_dev_network () ? std::chrono::seconds (2) : this_l->node.network_params.node.idle_timeout);
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_success, nano::stat::dir::in);
+				this_l->connections.push_back (new_connection);
+				callback_a (new_connection, ec_a);
+				this_l->on_connection (callback_a);
+				return;
+			}
+
+			this_l->node.logger.always_log ("Network: Unable to accept connection: ", ec_a.message ());
+			if (this_l->is_temporary_error (ec_a))
+			{
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
+				boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
+					this_l->on_connection (callback_a);
 				}));
+				return;
 			}
-		}
+
+			if (this_l->is_system_error (ec_a))
+			{
+				boost::this_thread::sleep (boost::posix_time::milliseconds (500));
+				this_l->node.stats.inc (nano::stat::type::tcp, nano::stat::detail::tcp_accept_failure, nano::stat::dir::in);
+				boost::asio::post (this_l->strand, boost::asio::bind_executor (this_l->strand, [this_l, callback_a] () {
+					this_l->on_connection (callback_a);
+				}));
+				return;
+			}
+
+			// Unhandled errors above
+			//EBADF
+			//EFAULT
+			//EINTR
+			//EINVAL
+			//ENOTSOCK
+			//EOPNOTSUPP
+			//EPROTO
+			//EPERM
+			// Check how the listener wants to handle these error
+			if (callback_a (new_connection, ec_a))
+			{
+				this_l->on_connection (callback_a);
+				return;
+			}
+			this_l->node.logger.always_log ("Network: Stopping to accept connections. Node may need restarting.");
+		}));
 	}));
 }
 
+bool nano::server_socket::is_temporary_error (boost::system::error_code const ec_a)
+{
+	switch (ec_a.value ())
+	{
+#if EAGAIN != EWOULDBLOCK
```

Comment:
Think no need to check here. Common practice is to check for both values, no matter the system.

Reply:
Results in compilation error on systems which match as a switch can't have the same value twice

Reply:
Ok. Using if statement would work. But would make the overall code more unclear I think, so agree this is better.